### ISSUE TITLE
Error 401 Unauthorized when trying to connect to appsync when using aws-amplify

### ISSUE THREAD

**Issue by codermeisje** [aws-amplify/amplify-js#4270]
** Which Category is your question related to? **
Using aws-amplify-vue, cognito, and appsync. Works great locally, but after it is deployed through AWS amplify console, I get a 401 Unauthorized error when trying to connect to the graphql endpoint.

** What AWS Services are you utilizing? **
Cognito, AppSync

** Provide additional details e.g. code snippets **
Components I am using from aws-amplify-vue:
- I am using the `<amplify-authenticator>` component for logging in
- I am using the `<amplify-connect>` component to run the graphql query

**Comment by tomgreenwood1** [comment]
I am getting the same error. I added hosting via the amplify cli. When I serve the website on localhost it can authenticate, when I run amplify publish it can authenticate, but when I merge it onto a 'deploy' branch and push via the amplify console, I get

`{
  "errors" : [ {
    "errorType" : "UnauthorizedException",
    "message" : "Missing authorization header"
  } ]
}`

from appsync. 

Also:  In the reponse header shows:

`x-amz-cf-id: oltBBX0aht9N2S52gRCm0PIzxxxxxxxxz9lqIFSTOKOCFD5e0Gcw==
x-amz-cf-pop: LHR4
x-amzn-errortype: UnauthorizedException
x-amzn-requestid: 4ecf8463-bea9-4a2f-xxxx-xxxx5875a842
x-cache: Error from cloudfront`

I can see in the request header, that the origin is:

`Origin: https://staging.myappid.amplifyapp.com`

but the origin on the cloudfront in the stack for that env is quite different.. is it that amplify console is getting confused that the backend already has a hosting element?

**Comment by tomgreenwood1** [comment]
I am confirm that EgidioCaprino's answer from #3089 fixed the problem for me. 

i.e. adding

`graphql_headers: async () => {
  const currentSession = await Auth.currentSession();
  return { Authorization: currentSession.getIdToken().getJwtToken() };
};`

into the awsconfig that's passed to Amplify.configure

**Comment by codermeisje** [comment]
That did the trick!  Thanks @tomgreenwood1

**Comment by marcuscodesok** [comment]
Newb but hopefully this is helpful:

I may have a different authentication method but I was getting a 401 error with GraphQL using AWS Amplify with a GraphQL API.

With my authentication method `./src/aws-exports.js` uses `API_KEY` as `aws_appsync_authenticationType`.

After some digging, I realized in `./amplify/backend/backend-config.json` the API key by default expires in seven days (something I thoughtlessly clicked through when spinning up the API with `amplify add api`). 

So I went about updating this. I'm sure there are command line solutions to updating the API key but I went into my AppySync console > API > name of my API > Settings and added a new API key. 

I then updated `./amplify/backend/backend-config.json` with the API description e.g., `            "apiKeyConfig": {
              "description": "yourNewApiKey"
            }`
and updated `./aws-exports.js` with the new API key `    "aws_appsync_apiKey": "newkeyinfo"
`.
